src/parser/metrics/RealLinesOfCode.ts

```diff
@@ -16,75 +12,75 @@ let dlog: DebugLoggerFunction = debuglog("metric-gardener", (logger) => {
-                }
+    /**
+     * Waling through the tree in order to find actual code lines.
+     *
+     * Uses a {@link TreeCursor} for this, as according to the
+     * {@link https://tree-sitter.github.io/tree-sitter/using-parsers#walking-trees-with-tree-cursors|Tree-sitter documentation},
+     * this is the most efficient way to traverse a syntax tree.
+     * @param cursor A {@link TreeCursor} for the syntax tree.
+     * @param sureCodeLines A set in which the line numbers of the found code lines are stored.
+     */
+    walkTree(cursor: TreeCursor, sureCodeLines: Set<number>) {
+        // This is not a comment syntax node, so assume it includes "real code".
+        if (!this.commentStatementsSet.has(cursor.currentNode.type)) {
+            // Languages like Go do use new line characters for marking the end of a statement, which leads to wrong
+            // results here, as they have an end position on the next line. So filter them out as well:
+            const isNewLineNode = /\n/.test(cursor.currentNode.type); // Should also filter \r\n, etc.
+            if (!isNewLineNode) {
+                // Assume that first and last line of whatever kind of node this is, is a real code line.
+                // This assumption should hold for all kinds of block/composed statements in (hopefully) all languages.
+                sureCodeLines.add(cursor.startPosition.row);
+                // sureCodeLines.add(cursor.endPosition.row);
+                // Adding the last line is not necessary, as every last line has to have some syntactical element,
+                // which is again expressed as another syntax node.
             }
-        });
+        }
+        // Recurse, depth-first
+        if (cursor.gotoFirstChild()) {
+            this.walkTree(cursor, sureCodeLines);
+        }
+        if (cursor.gotoNextSibling()) {
+            this.walkTree(cursor, sureCodeLines);
+        } else {
+            // Completed searching this part of the tree, so go up now.
+            cursor.gotoParent();
+        }
     }
 
     calculate(parseFile: ParseFile): MetricResult {
         const tree = TreeParser.getParseTree(parseFile);
+        const sureCodeLines = new Set<number>();
 
-        // Avoid off-by-one error:
-        // The number of the last row equals the number of lines in the file minus one,
-        // as it is counted from line 0. So add one to the result:
-        const loc = tree.rootNode.endPosition.row + 1;
-
-        const emptyLines = this.countEmptyLines(tree.rootNode.text);
-
-        const queryBuilder = new QueryBuilder(
-            grammars.get(parseFile.language),
-            tree,
-            parseFile.language
-        );
-
-        queryBuilder.setStatements(this.commentStatementsSuperSet);
+        const cursor = tree.walk();
+        // Assume the root node is always some kind of program/file/compilation_unit stuff
+        cursor.gotoFirstChild();
+        this.walkTree(cursor, sureCodeLines);
```

Comment:
What about letting the first method call to set the set variable and to let the method return that?
That way we do not change an outside variable inside the method.

```suggestion
       const actualCodeLines = this.walkTree(cursor);
```

Reply:
So I should introduce something like the following?
`function walkTree1(cursor): Set<number>{
    let sureCodeLines = new Set();
    this.walkTree(cursor, sureCodeLines);
    return sureCodeLines;
}`

Otherwise I would have to instantiate a new set with every recursive call and merge that after returning from each recursive call, which I guess will be quite inefficient compared to now?

Reply:
I meant something like this:

```js
function walkTree(cursor, sureCodeLines = new Set()) {
  // regular walkTree logic
  return sureCodeLines;
}
```

Reply:
Ah, yes, that looks more elegant. I have changed the code accordingly.